lib/atmega328/max72xx.cpp

```diff
@@ -1,70 +1,75 @@
 #include "max72xx.h"
-#include "util.h"
 #include <avr/io.h>
 #include <util/delay.h>
+#include "util.h"
+#include "SPI.h"
 
-#define PD_CLK PD5
-#define PD_CS PD6
-#define PD_DATA PD7
+#define CS_PIN PB2
 
-static void inline deactivate_cs(void);
-static void inline activate_cs(void);
-static void assert_clock(void);
-static void inline deassert_clock(void);
+void max72xx_send_commands(max72xx_cmd_t *cmds, uint8_t length);
+void max72xx_send_commands_to_all(max72xx_reg_t reg, uint8_t data);
 
 void init_max72xx(void)
 {
-    PORTD &= ~(bit(PD_CLK) | bit(PD_DATA));
-    deactivate_cs();
-    DDRD |= bit(PD_CLK) | bit(PD_CS) | bit(PD_DATA);
+    init_SPI(CS_PIN);
+
+    max72xx_send_commands_to_all(Max72XX_Shutdown, 0x01);        // normal operation (exit shutdown mode)
+    max72xx_send_commands_to_all(Max72XX_Scan_Limit, 0x07);      // 8 digits scan limit
+    max72xx_send_commands_to_all(Max72XX_Decode_Mode, 0x00);     // disable decode mode
+    max72xx_send_commands_to_all(Max72XX_Intensity, 0x0F);       // set brightness
+    max72xx_send_commands_to_all(Max72XX_Display_Test, 0x00);    // disable display test
 }
 
-static void send_byte(uint8_t data)
+static void inline deactivate_cs(void)
 {
-    for (int8_t i = 7; i >= 0; i--)
-    {
-        bool bit_at_offset_i_is_set = bit(7) & data;
-        uint8_t b = PORTD & ~bit(PD_DATA);
-        PORTD = b | (bit_at_offset_i_is_set ? bit(PD_DATA) : 0);
-        _delay_us(1);
-
-        assert_clock();
-        _delay_us(1);
+    PORTB |= bit(CS_PIN);
+}
 
-        deassert_clock();
-        data <<= 1;
-    }
+static void inline activate_cs(void)
+{
+    PORTB &= ~bit(CS_PIN);
 }
 
+
 void max72xx_send_commands(max72xx_cmd_t *cmds, uint8_t length)
 {
     activate_cs();
-
-    for (uint8_t i = 0; i < length; i++)
+    for (uint8_t device = 0; device < length; device++)
     {
-        send_byte((uint8_t)cmds[i].reg);
-        send_byte(cmds[i].data);
+        SPI_transmit_byte(cmds[device].reg);
+        SPI_transmit_byte(cmds[device].data);
     }
-
     deactivate_cs();
 }
 
-static void inline deactivate_cs(void)
+void max72xx_send_commands_to_all(max72xx_reg_t reg, uint8_t data)
 {
-    PORTD |= bit(PD_CS);
+    max72xx_cmd_t cmds[MAX72XX_NUM_DEVICES];
+    for (int i=0; i < MAX72XX_NUM_DEVICES; i++)
+    {
+        cmds[i].reg = reg;
+        cmds[i].data = data;
+    }
+    max72xx_send_commands(cmds, MAX72XX_NUM_DEVICES);
 }
 
-static void inline activate_cs(void)
+void max72xx_clear(max72xx_reg_t row)
 {
-    PORTD &= ~bit(PD_CS);
+    max72xx_send_commands_to_all(row, 0x00);
```

Comment:
Maybe easier to read?
```suggestion
     max72xx_send_commands_to_all(Max72XX_Digit0 + row, 0x00);
```